src/Pact/Analyze/Eval/Term.hs

```diff
@@ -344,10 +344,51 @@ writeFields writeType tid tn sRk (S mProv obj)
 
 writeFields _ _ _ _ _ _ = vacuousMatch "the previous two cases are complete"
 
+accumulatingPendingGrants :: Analyze () -> Analyze CapabilityGrants
+accumulatingPendingGrants act = do
+    emptyGrants <- view aeEmptyGrants
+    previouslyPending <- swapIn emptyGrants
+    act
+    swapIn previouslyPending
+
+  where
+    swapIn :: CapabilityGrants -> Analyze CapabilityGrants
+    swapIn next = do
+      prev <- use $ latticeState.lasPendingGrants
+      latticeState.lasPendingGrants .= next
+      pure prev
+
+capabilityAppToken :: Capability -> [VarId] -> Analyze Token
+capabilityAppToken (Capability schema capName) vids = do
+    mAVals <- sequence <$> traverse getVar vids
+    case mAVals of
+      Nothing    ->
+        throwErrorNoLoc "unexpected missing variable in environment"
+      Just avals ->
+        pure $ Token schema capName $ sansProv $ buildObject schema avals
+
+  where
+    buildObject :: SingList schema -> [AVal] -> SBV (ConcreteObj schema)
+    buildObject SNil' [] = literal ()
```

Comment:
Prefer `foldSingList` or `foldrSingList`. I think the `foldr` version should work here.

Reply:
Note that I am recursing over the two structures in parallel. From what I understand, I _could_ use `foldrHList` if there were an easy way to go from `[AVal]` to an `HList`, and then if there were a combinator to zip the `HLists`. This seemed more straightforward to me at the time. LMK if you have a suggestion from here.

Reply:
I decided this is the best way to do it